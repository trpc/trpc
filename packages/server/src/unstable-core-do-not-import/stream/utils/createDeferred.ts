/* eslint-disable @typescript-eslint/no-non-null-assertion */
export function createDeferred<TValue>() {
  let resolve: (value: TValue) => void;
  let reject: (error: unknown) => void;
  const promise = new Promise<TValue>((res, rej) => {
    resolve = res;
    reject = rej;
  });

  return { promise, resolve: resolve!, reject: reject! };
}
export type Deferred<TValue> = ReturnType<typeof createDeferred<TValue>>;

export const createTimeoutPromise = <TValue>(
  timeoutMs: number,
  value: TValue,
) => {
  let deferred = createDeferred<TValue>();
  deferred = deferred as typeof deferred & { clear: () => void };

  let timeout: ReturnType<typeof setTimeout> | null = null;
  const clear = () => {
    if (timeout) {
      clearTimeout(timeout);
      timeout = null;
    }
  };
  const resolve = () => {
    deferred.resolve(value);
    clear();
  };
  if (timeoutMs !== Infinity) {
    timeout = setTimeout(resolve, timeoutMs);
    timeout.unref?.();
  }

  return {
    promise: deferred.promise,
    /**
     * Clear the timeout without resolving the promise
     */
    clear,
    /**
     * Resolve the promise with the value
     */
    resolve,
  };
};
