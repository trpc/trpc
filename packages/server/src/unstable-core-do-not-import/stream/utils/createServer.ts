import http from 'http';
import type { Socket } from 'net';

export function createServer(
  handler: (req: http.IncomingMessage, res: http.ServerResponse) => void,
) {
  const server = http.createServer(async (req, res) => {
    handler(req, res);
  });
  server.listen(0);

  const connections = new Set<Socket>();
  server.on('connection', (conn) => {
    connections.add(conn);
    conn.once('close', () => {
      connections.delete(conn);
    });
  });

  const port = (server.address() as any).port;

  const url = `http://localhost:${port}`;

  async function forceClose() {
    for (const conn of connections) {
      conn.destroy();
    }
    await new Promise((resolve) => {
      server.close(resolve);
    });
  }

  return {
    url,
    close: async () => {
      await forceClose();
    },
    restart: async () => {
      await forceClose();

      server.listen(port);
    },
  };
}
