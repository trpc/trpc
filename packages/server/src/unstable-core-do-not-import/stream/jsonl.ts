import { Unpromise } from '../../vendor/unpromise';
import { getTRPCErrorFromUnknown } from '../error/TRPCError';
import { isAsyncIterable, isFunction, isObject, run } from '../utils';
import type { Deferred } from './utils/createDeferred';
import { createDeferred } from './utils/createDeferred';
import {
  createReadableStream,
  isCancelledStreamResult,
} from './utils/createReadableStream';

/**
 * A subset of the standard ReadableStream properties needed by tRPC internally.
 * @see ReadableStream from lib.dom.d.ts
 */
export type WebReadableStreamEsque = {
  getReader: () => ReadableStreamDefaultReader<Uint8Array>;
};

export type NodeJSReadableStreamEsque = {
  on(
    eventName: string | symbol,
    listener: (...args: any[]) => void,
  ): NodeJSReadableStreamEsque;
};

// ---------- types
const CHUNK_VALUE_TYPE_PROMISE = 0;
type CHUNK_VALUE_TYPE_PROMISE = typeof CHUNK_VALUE_TYPE_PROMISE;
const CHUNK_VALUE_TYPE_ASYNC_ITERABLE = 1;
type CHUNK_VALUE_TYPE_ASYNC_ITERABLE = typeof CHUNK_VALUE_TYPE_ASYNC_ITERABLE;

const PROMISE_STATUS_FULFILLED = 0;
type PROMISE_STATUS_FULFILLED = typeof PROMISE_STATUS_FULFILLED;
const PROMISE_STATUS_REJECTED = 1;
type PROMISE_STATUS_REJECTED = typeof PROMISE_STATUS_REJECTED;

const ASYNC_ITERABLE_STATUS_RETURN = 0;
type ASYNC_ITERABLE_STATUS_RETURN = typeof ASYNC_ITERABLE_STATUS_RETURN;
const ASYNC_ITERABLE_STATUS_VALUE = 1;
type ASYNC_ITERABLE_STATUS_VALUE = typeof ASYNC_ITERABLE_STATUS_VALUE;
const ASYNC_ITERABLE_STATUS_ERROR = 2;
type ASYNC_ITERABLE_STATUS_ERROR = typeof ASYNC_ITERABLE_STATUS_ERROR;

type ChunkDefinitionKey =
  // root should be replaced
  | null
  // at array path
  | number
  // at key path
  | string;

type ChunkIndex = number & { __chunkIndex: true };
type ChunkValueType =
  | CHUNK_VALUE_TYPE_PROMISE
  | CHUNK_VALUE_TYPE_ASYNC_ITERABLE;
type ChunkDefinition = [
  key: ChunkDefinitionKey,
  type: ChunkValueType,
  chunkId: ChunkIndex,
];
type DehydratedValue = [
  // data
  [unknown] | [],
  // chunk descriptions
  ...ChunkDefinition[],
];

type Head = Record<string, DehydratedValue>;
type PromiseChunk =
  | [
      chunkIndex: ChunkIndex,
      status: PROMISE_STATUS_FULFILLED,
      value: DehydratedValue,
    ]
  | [chunkIndex: ChunkIndex, status: PROMISE_STATUS_REJECTED, error: unknown];
type IterableChunk =
  | [
      chunkIndex: ChunkIndex,
      status: ASYNC_ITERABLE_STATUS_RETURN,
      value: DehydratedValue,
    ]
  | [
      chunkIndex: ChunkIndex,
      status: ASYNC_ITERABLE_STATUS_VALUE,
      value: DehydratedValue,
    ]
  | [
      chunkIndex: ChunkIndex,
      status: ASYNC_ITERABLE_STATUS_ERROR,
      error: unknown,
    ];
type ChunkData = PromiseChunk | IterableChunk;
type PlaceholderValue = 0 & { __placeholder: true };
export function isPromise(value: unknown): value is Promise<unknown> {
  return (
    (isObject(value) || isFunction(value)) &&
    typeof value?.['then'] === 'function' &&
    typeof value?.['catch'] === 'function'
  );
}

type Serialize = (value: any) => any;
type Deserialize = (value: any) => any;

type PathArray = readonly (string | number)[];
export type ProducerOnError = (opts: {
  error: unknown;
  path: PathArray;
}) => void;
export interface ProducerOptions {
  serialize?: Serialize;
  data: Record<string, unknown> | unknown[];
  onError?: ProducerOnError;
  formatError?: (opts: { error: unknown; path: PathArray }) => unknown;
  maxDepth?: number;
}

class MaxDepthError extends Error {
  constructor(public path: (string | number)[]) {
    super('Max depth reached at path: ' + path.join('.'));
  }
}

function createBatchStreamProducer(opts: ProducerOptions) {
  const { data } = opts;
  let counter = 0 as ChunkIndex;
  const placeholder = 0 as PlaceholderValue;

  const stream = createReadableStream<ChunkData>();
  const pending = new Set<ChunkIndex>();

  function maybeClose() {
    if (pending.size === 0 && !stream.cancelled()) {
      stream.controller.close();
    }
  }
  function dehydratePromise(
    promise: Promise<unknown>,
    path: (string | number)[],
  ) {
    //
    const error = checkMaxDepth(path);
    if (error) {
      promise.catch(() => {
        // ignore
      });
      promise = Promise.reject(error);
    }
    const idx = counter++ as ChunkIndex;
    pending.add(idx);

    Unpromise.race([promise, stream.cancelledPromise])
      .then((it) => {
        if (isCancelledStreamResult(it)) {
          return;
        }
        stream.controller.enqueue([
          idx,
          PROMISE_STATUS_FULFILLED,
          dehydrate(it, path),
        ]);
      })
      .catch((cause) => {
        opts.onError?.({ error: cause, path });
        stream.controller.enqueue([
          idx,
          PROMISE_STATUS_REJECTED,
          opts.formatError?.({ error: cause, path }),
        ]);
      })
      .finally(() => {
        pending.delete(idx);
        maybeClose();
      });
    return idx;
  }
  function dehydrateAsyncIterable(
    iterable: AsyncIterable<unknown>,
    path: (string | number)[],
  ) {
    const error = checkMaxDepth(path);
    if (error) {
      iterable = {
        [Symbol.asyncIterator]() {
          throw error;
        },
      };
    }
    const idx = counter++ as ChunkIndex;
    pending.add(idx);
    run(async () => {
      const iterator = iterable[Symbol.asyncIterator]();

      while (true) {
        const next = await Unpromise.race([
          iterator.next().catch(getTRPCErrorFromUnknown),
          stream.cancelledPromise,
        ]);

        if (next instanceof Error) {
          opts.onError?.({ error: next, path });

          stream.controller.enqueue([
            idx,
            ASYNC_ITERABLE_STATUS_ERROR,
            opts.formatError?.({ error: next, path }),
          ]);
          return;
        }
        if (isCancelledStreamResult(next)) {
          await iterator.return?.();
          break;
        }
        if (next.done) {
          stream.controller.enqueue([
            idx,
            ASYNC_ITERABLE_STATUS_RETURN,
            dehydrate(next.value, path),
          ]);
          break;
        }
        stream.controller.enqueue([
          idx,
          ASYNC_ITERABLE_STATUS_VALUE,
          dehydrate(next.value, path),
        ]);
      }

      pending.delete(idx);
      maybeClose();
    }).catch((cause) => {
      // this shouldn't happen, but node crashes if we don't catch it
      opts.onError?.({
        error: new Error(
          'You found a bug - please report it on https://github.com/trpc/trpc',
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore https://github.com/tc39/proposal-error-cause
          {
            cause,
          },
        ),
        path,
      });
    });
    return idx;
  }
  function checkMaxDepth(path: (string | number)[]) {
    if (opts.maxDepth && path.length > opts.maxDepth) {
      return new MaxDepthError(path);
    }
    return null;
  }
  function dehydrateAsync(
    value: unknown,
    path: (string | number)[],
  ): null | [type: ChunkValueType, chunkId: ChunkIndex] {
    if (isPromise(value)) {
      return [CHUNK_VALUE_TYPE_PROMISE, dehydratePromise(value, path)];
    }
    if (isAsyncIterable(value)) {
      if (opts.maxDepth && path.length >= opts.maxDepth) {
        throw new Error('Max depth reached');
      }
      return [
        CHUNK_VALUE_TYPE_ASYNC_ITERABLE,
        dehydrateAsyncIterable(value, path),
      ];
    }
    return null;
  }
  function dehydrate(
    value: unknown,
    path: (string | number)[],
  ): DehydratedValue {
    if (value === undefined) {
      return [[]];
    }
    if (!isObject(value)) {
      return [[value]];
    }
    const reg = dehydrateAsync(value, path);
    if (reg) {
      return [[placeholder], [null, ...reg]];
    }
    const newObj = {} as Record<string, unknown>;
    const asyncValues: ChunkDefinition[] = [];
    for (const [key, item] of Object.entries(value)) {
      const transformed = dehydrateAsync(item, [...path, key]);
      if (!transformed) {
        newObj[key] = item;
        continue;
      }
      newObj[key] = placeholder;
      asyncValues.push([key, ...transformed]);
    }
    return [[newObj], ...asyncValues];
  }

  const newHead: Head = {};
  for (const [key, item] of Object.entries(data)) {
    newHead[key] = dehydrate(item, [key]);
  }

  return [newHead, stream.readable] as const;
}
/**
 * JSON Lines stream producer
 * @see https://jsonlines.org/
 */
export function jsonlStreamProducer(opts: ProducerOptions) {
  let [head, stream] = createBatchStreamProducer(opts);

  const { serialize } = opts;
  if (serialize) {
    head = serialize(head);
    stream = stream.pipeThrough(
      new TransformStream({
        transform(chunk, controller) {
          controller.enqueue(serialize(chunk));
        },
      }),
    );
  }

  return stream
    .pipeThrough(
      new TransformStream({
        start(controller) {
          controller.enqueue(JSON.stringify(head) + '\n');
        },
        transform(chunk, controller) {
          controller.enqueue(JSON.stringify(chunk) + '\n');
        },
      }),
    )
    .pipeThrough(new TextEncoderStream());
}

class StreamInterruptedError extends Error {
  constructor(cause?: unknown) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore https://github.com/tc39/proposal-error-cause
    super('Invalid response or stream interrupted', { cause });
  }
}
class AsyncError extends Error {
  constructor(public readonly data: unknown) {
    super('Received error from server');
  }
}
export type ConsumerOnError = (opts: { error: unknown }) => void;

const nodeJsStreamToReaderEsque = (source: NodeJSReadableStreamEsque) => {
  return {
    getReader() {
      const { readable, controller } = createReadableStream<Uint8Array>();
      source.on('data', (chunk) => {
        controller.enqueue(chunk);
      });
      source.on('end', () => {
        controller.close();
      });
      source.on('error', (error) => {
        controller.error(error);
      });
      return readable.getReader();
    },
  };
};

function createLineAccumulator(
  from: NodeJSReadableStreamEsque | WebReadableStreamEsque,
) {
  const reader =
    'getReader' in from
      ? from.getReader()
      : nodeJsStreamToReaderEsque(from).getReader();

  let lineAggregate = '';

  return new ReadableStream({
    async pull(controller) {
      const { done, value } = await reader.read();

      if (done) {
        controller.close();
      } else {
        controller.enqueue(value);
      }
    },
    cancel() {
      return reader.cancel();
    },
  })
    .pipeThrough(new TextDecoderStream())
    .pipeThrough(
      new TransformStream<string, string>({
        transform(chunk, controller) {
          lineAggregate += chunk;
          const parts = lineAggregate.split('\n');
          lineAggregate = parts.pop() ?? '';
          for (const part of parts) {
            controller.enqueue(part);
          }
        },
      }),
    );
}
function createConsumerStream<THead>(
  from: NodeJSReadableStreamEsque | WebReadableStreamEsque,
) {
  const stream = createLineAccumulator(from);

  let sentHead = false;
  return stream.pipeThrough(
    new TransformStream<string, ChunkData | THead>({
      transform(line, controller) {
        if (!sentHead) {
          const head = JSON.parse(line);
          controller.enqueue(head as THead);
          sentHead = true;
        } else {
          const chunk: ChunkData = JSON.parse(line);
          controller.enqueue(chunk);
        }
      },
    }),
  );
}

/**
 * JSON Lines stream consumer
 * @see https://jsonlines.org/
 */
export async function jsonlStreamConsumer<THead>(opts: {
  from: NodeJSReadableStreamEsque | WebReadableStreamEsque;
  deserialize?: Deserialize;
  onError?: ConsumerOnError;
  formatError?: (opts: { error: unknown }) => Error;
  /**
   * This `AbortController` will be triggered when there are no more listeners to the stream.
   */
  abortController: AbortController | null;
}) {
  const { deserialize = (v) => v } = opts;

  let source = createConsumerStream<Head>(opts.from);
  if (deserialize) {
    source = source.pipeThrough(
      new TransformStream({
        transform(chunk, controller) {
          controller.enqueue(deserialize(chunk));
        },
      }),
    );
  }
  let headDeferred: null | Deferred<THead> = createDeferred();

  type ControllerChunk = ChunkData | StreamInterruptedError;
  type ChunkController = ReadableStreamDefaultController<ControllerChunk>;
  type ControllerWrapper = {
    controller: ChunkController;
    returned: boolean;
  };
  const chunkDeferred = new Map<ChunkIndex, Deferred<ControllerWrapper>>();
  const controllers = new Map<ChunkIndex, ControllerWrapper>();

  const maybeAbort = () => {
    if (
      chunkDeferred.size === 0 &&
      Array.from(controllers.values()).every((it) => it.returned)
    ) {
      // nothing is listening to the stream anymore
      opts.abortController?.abort();
    }
  };

  function hydrateChunkDefinition(value: ChunkDefinition) {
    const [_path, type, chunkId] = value;

    const { readable, controller } = createReadableStream<ChunkData>();

    const wrapper: ControllerWrapper = {
      controller,
      returned: false,
    };
    controllers.set(chunkId, wrapper);

    // resolve chunk deferred if it exists
    const deferred = chunkDeferred.get(chunkId);
    if (deferred) {
      deferred.resolve(wrapper);
      chunkDeferred.delete(chunkId);
    }

    switch (type) {
      case CHUNK_VALUE_TYPE_PROMISE: {
        return new Promise((resolve, reject) => {
          // listen for next value in the stream
          const reader = readable.getReader();
          reader
            .read()
            .then((it) => {
              if (it.done) {
                reject(new Error('Promise chunk ended without value'));
                return;
              }
              if (it.value instanceof StreamInterruptedError) {
                reject(it.value);
                return;
              }
              const value = it.value;
              const [_chunkId, status, data] = value as PromiseChunk;
              switch (status) {
                case PROMISE_STATUS_FULFILLED:
                  resolve(hydrate(data));

                  break;
                case PROMISE_STATUS_REJECTED:
                  reject(
                    opts.formatError?.({ error: data }) ?? new AsyncError(data),
                  );
                  break;
              }
            })
            .catch(reject)
            .finally(() => {
              // reader.releaseLock();
              controllers.delete(chunkId);

              maybeAbort();
            });
        });
      }
      case CHUNK_VALUE_TYPE_ASYNC_ITERABLE: {
        return {
          [Symbol.asyncIterator]: () => {
            const reader = readable.getReader();
            const iterator: AsyncIterator<unknown> = {
              next: async () => {
                const { done, value } = await reader.read();
                if (value instanceof StreamInterruptedError) {
                  throw value;
                }
                if (done) {
                  controllers.delete(chunkId);
                  maybeAbort();
                  return {
                    done: true,
                    value: undefined,
                  };
                }

                const [_chunkId, status, data] = value as IterableChunk;

                switch (status) {
                  case ASYNC_ITERABLE_STATUS_VALUE:
                    return {
                      done: false,
                      value: hydrate(data),
                    };
                  case ASYNC_ITERABLE_STATUS_RETURN:
                    controllers.delete(chunkId);
                    maybeAbort();
                    return {
                      done: true,
                      value: hydrate(data),
                    };
                  case ASYNC_ITERABLE_STATUS_ERROR:
                    controllers.delete(chunkId);
                    maybeAbort();
                    throw (
                      opts.formatError?.({ error: data }) ??
                      new AsyncError(data)
                    );
                }
              },
              return: async () => {
                wrapper.returned = true;
                maybeAbort();
                return {
                  done: true,
                  value: undefined,
                };
              },
            };
            return iterator;
          },
        };
      }
    }
  }

  function hydrate(value: DehydratedValue): unknown {
    const [[data], ...asyncProps] = value;

    for (const value of asyncProps) {
      const [key] = value;
      const hydrated = hydrateChunkDefinition(value);

      if (key === null) {
        return hydrated;
      }

      (data as any)[key] = hydrated;
    }
    return data;
  }

  const closeOrAbort = (reason?: unknown) => {
    const error = new StreamInterruptedError(reason);

    headDeferred?.reject(error);
    for (const deferred of chunkDeferred.values()) {
      deferred.reject(error);
    }
    chunkDeferred.clear();
    for (const { controller } of controllers.values()) {
      controller.enqueue(error);
      controller.close();
    }
    controllers.clear();
  };
  source
    .pipeTo(
      new WritableStream({
        async write(chunkOrHead) {
          if (headDeferred) {
            const head = chunkOrHead as Record<number | string, unknown>;

            for (const [key, value] of Object.entries(chunkOrHead)) {
              const parsed = hydrate(value as any);
              head[key] = parsed;
            }
            headDeferred.resolve(head as THead);
            headDeferred = null;
            return;
          }
          const chunk = chunkOrHead as ChunkData;
          const [idx] = chunk;
          let wrapper = controllers.get(idx);
          if (!wrapper) {
            let deferred = chunkDeferred.get(idx);
            if (!deferred) {
              deferred = createDeferred();
              chunkDeferred.set(idx, deferred);
            }

            wrapper = await deferred.promise;
          }
          wrapper.controller.enqueue(chunk);
        },
        close: closeOrAbort,
        abort: closeOrAbort,
      }),
    )
    .catch((error) => {
      opts.onError?.({ error });
      closeOrAbort(error);
    });

  return [
    await headDeferred.promise,
    {
      controllers,
    },
  ] as const;
}
