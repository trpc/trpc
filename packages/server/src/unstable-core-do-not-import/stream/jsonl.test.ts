import { EventEmitter, on } from 'node:events';
import { waitFor } from '@testing-library/react';
import SuperJSON from 'superjson';
import type { ConsumerOnError, ProducerOnError } from './jsonl';
import { jsonlStreamConsumer, jsonlStreamProducer } from './jsonl';
import { createServer } from './utils/createServer';

test('encode/decode with superjson', async () => {
  const data = {
    0: Promise.resolve({
      foo: {
        bar: {
          baz: 'qux',
        },
      },
      deferred: Promise.resolve(42),
    }),
    1: Promise.resolve({
      [Symbol.asyncIterator]: async function* () {
        yield 1;
        yield 2;
        yield 3;
      },
    }),
  } as const;
  const stream = jsonlStreamProducer({
    data,
    serialize: (v) => SuperJSON.serialize(v),
  });

  const [head, meta] = await jsonlStreamConsumer<typeof data>({
    from: stream,
    deserialize: (v) => SuperJSON.deserialize(v),
    abortController: null,
  });

  {
    expect(head[0]).toBeInstanceOf(Promise);

    const value = await head[0];
    expect(value.deferred).toBeInstanceOf(Promise);

    await expect(value.deferred).resolves.toBe(42);

    expect(value.foo.bar.baz).toBe('qux');
  }
  {
    expect(head[1]).toBeInstanceOf(Promise);

    const iterable = await head[1];

    const aggregated: number[] = [];
    for await (const item of iterable) {
      aggregated.push(item);
    }
    expect(aggregated).toEqual([1, 2, 3]);
  }

  expect(meta.controllers.size).toBe(0);
});

test('encode/decode - error', async () => {
  const data = {
    0: Promise.resolve({
      foo: {
        bar: {
          baz: 'qux',
        },
      },
      deferred: Promise.reject(new Error('promise')),
    }),
    1: Promise.resolve({
      [Symbol.asyncIterator]: async function* () {
        yield 1;
        yield 2;
        yield 3;
        throw new Error('iterable');
      },
    }),
  } as const;

  const errors: unknown[] = [];

  const onProducerErrorSpy =
    vi.fn<(...args: Parameters<ProducerOnError>) => void>();
  const onConsumerErrorSpy =
    vi.fn<(...args: Parameters<ConsumerOnError>) => void>();

  const stream = jsonlStreamProducer({
    data,
    serialize: (v) => SuperJSON.serialize(v),
    onError: onProducerErrorSpy,
  });

  const [head, meta] = await jsonlStreamConsumer<typeof data>({
    from: stream,
    deserialize: (v) => SuperJSON.deserialize(v),
    onError: onConsumerErrorSpy,
    abortController: null,
  });

  {
    expect(head[0]).toBeInstanceOf(Promise);

    const value = await head[0];
    expect(value.deferred).toBeInstanceOf(Promise);

    await expect(value.deferred).rejects.toThrowErrorMatchingInlineSnapshot(
      `[Error: Received error from server]`,
    );
  }
  {
    expect(head[1]).toBeInstanceOf(Promise);

    const iterable = await head[1];

    const aggregated: number[] = [];
    try {
      for await (const item of iterable) {
        aggregated.push(item);
      }
    } catch (err) {
      errors.push(err);
    }
    expect(aggregated).toEqual([1, 2, 3]);

    expect(errors).toHaveLength(1);
    expect(errors[0]).toMatchInlineSnapshot(
      `[Error: Received error from server]`,
    );
  }

  expect(onProducerErrorSpy).toHaveBeenCalledTimes(2);
  expect(onProducerErrorSpy.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "error": [Error: promise],
          "path": Array [
            "0",
            "deferred",
          ],
        },
      ],
      Array [
        Object {
          "error": [TRPCError: iterable],
          "path": Array [
            "1",
          ],
        },
      ],
    ]
  `);

  // await meta.reader.closed;
  expect(meta.controllers.size).toBe(0);
  expect(onConsumerErrorSpy).toHaveBeenCalledTimes(0);
});

test('decode - bad data', async () => {
  const textEncoder = new TextEncoderStream();
  const writer = textEncoder.writable.getWriter();

  try {
    (async () => {
      await writer.write(
        JSON.stringify({
          error: 'bad data',
        }),
      );
      await writer.close();
    })().catch(() => {
      // noop
    });
    await jsonlStreamConsumer({
      from: textEncoder.readable,
      deserialize: (v) => SuperJSON.deserialize(v),
      abortController: null,
    });
    expect(true).toBe(false);
  } catch (err) {
    expect(err).toMatchInlineSnapshot(
      `[Error: Invalid response or stream interrupted]`,
    );
  }
});

function createServerForStream(
  stream: ReadableStream,
  abortSignal: AbortController,
  headers: Record<string, string> = {},
) {
  return createServer(async (req, res) => {
    req.once('aborted', () => {
      abortSignal.abort();
    });
    for (const [key, value] of Object.entries(headers)) {
      res.setHeader(key, value);
    }

    const reader = stream.getReader();
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }

      res.write(value);
    }
    res.end();
  });
}
test('e2e, create server', async () => {
  const data = {
    0: Promise.resolve({
      foo: {
        bar: {
          baz: 'qux',
        },
      },
      deferred: Promise.resolve(42),
      slow: new Promise<string>((resolve) => {
        setTimeout(() => {
          resolve('___________RESOLVE________');
        }, 10);
      }),
    }),
    1: Promise.resolve({
      [Symbol.asyncIterator]: async function* () {
        yield 1;
        yield 2;
        yield 3;
      },
    }),
  } as const;
  const stream = jsonlStreamProducer({
    data,
    serialize: (v) => SuperJSON.serialize(v),
  });

  const server = createServerForStream(stream, new AbortController());

  const res = await fetch(server.url);

  const [head, meta] = await jsonlStreamConsumer<typeof data>({
    from: res.body!,
    deserialize: (v) => SuperJSON.deserialize(v),
    abortController: null,
  });

  {
    expect(head[0]).toBeInstanceOf(Promise);

    const value = await head[0];
    expect(value.deferred).toBeInstanceOf(Promise);

    await expect(value.deferred).resolves.toBe(42);

    expect(value.foo.bar.baz).toBe('qux');
  }
  {
    expect(head[1]).toBeInstanceOf(Promise);

    const iterable = await head[1];

    const aggregated: number[] = [];
    for await (const item of iterable) {
      aggregated.push(item);
    }
    expect(aggregated).toEqual([1, 2, 3]);
  }

  {
    const value = await head[0];
    expect(value.slow).toBeInstanceOf(Promise);
    await expect(value.slow).resolves.toMatchInlineSnapshot(
      `"___________RESOLVE________"`,
    );
  }
  // await meta.reader.closed;
  expect(meta.controllers.size).toBe(0);

  await server.close();
});

test('e2e, client aborts request halfway through', async () => {
  const serverAbort = new AbortController();
  const clientAbort = new AbortController();
  const yieldCalls = vi.fn();
  let stopped = false;

  const onConsumerErrorSpy =
    vi.fn<(...args: Parameters<ConsumerOnError>) => void>();
  const onProducerErrorSpy =
    vi.fn<(...args: Parameters<ProducerOnError>) => void>();

  const data = {
    0: Promise.resolve({
      [Symbol.asyncIterator]: async function* () {
        for (let i = 0; i < 10; i++) {
          yieldCalls();
          yield i;
          await new Promise((resolve) => setTimeout(resolve, 5));

          if (serverAbort.signal.aborted) {
            stopped = true;
            return;
          }
        }
        stopped = true;
      },
    }),
  } as const;

  const stream = jsonlStreamProducer({
    data,
    onError: onProducerErrorSpy,
  });
  const server = createServerForStream(stream, serverAbort);

  const res = await fetch(server.url, {
    signal: clientAbort.signal,
  });
  const [head, meta] = await jsonlStreamConsumer<typeof data>({
    from: res.body!,
    onError: onConsumerErrorSpy,
    abortController: clientAbort,
  });

  {
    const iterable = await head[0];

    for await (const item of iterable) {
      if (item === 2) {
        break;
      }
    }
    clientAbort.abort();
  }

  await waitFor(() => {
    expect(meta.controllers.size).toBe(0);
  });
  // wait for stopped
  await waitFor(() => {
    expect(stopped).toBe(true);
  });

  expect(yieldCalls.mock.calls.length).toBeGreaterThanOrEqual(3);
  expect(yieldCalls.mock.calls.length).toBeLessThan(10);

  const errors = onConsumerErrorSpy.mock.calls.map(
    (it) => (it[0].error as Error).message,
  );
  expect(errors).toMatchInlineSnapshot(`
    Array [
      "The operation was aborted.",
    ]
  `);
  expect(onConsumerErrorSpy).toHaveBeenCalledTimes(1);
  expect(onProducerErrorSpy).toHaveBeenCalledTimes(0);

  await server.close();
});

test('e2e, client aborts request halfway through - through breaking async iterable', async () => {
  const serverAbort = new AbortController();
  const clientAbort = new AbortController();
  const yieldCalls = vi.fn();
  let stopped = false;

  const onConsumerErrorSpy =
    vi.fn<(...args: Parameters<ConsumerOnError>) => void>();
  const onProducerErrorSpy =
    vi.fn<(...args: Parameters<ProducerOnError>) => void>();

  const data = {
    0: Promise.resolve({
      [Symbol.asyncIterator]: async function* () {
        for (let i = 0; i < 10; i++) {
          yieldCalls();
          yield i;
          await new Promise((resolve) => setTimeout(resolve, 5));
          if (serverAbort.signal.aborted) {
            stopped = true;
            return;
          }
        }
        stopped = true;
      },
    }),
  } as const;

  const stream = jsonlStreamProducer({
    data,
    onError: onProducerErrorSpy,
  });
  const server = createServerForStream(stream, serverAbort);

  const res = await fetch(server.url, {
    signal: clientAbort.signal,
  });
  const [head, meta] = await jsonlStreamConsumer<typeof data>({
    from: res.body!,
    onError: onConsumerErrorSpy,
    abortController: clientAbort,
  });

  {
    const iterable = await head[0];

    for await (const item of iterable) {
      if (item === 2) {
        // âœ¨ This will actually abort the full stream and the request since there's no more data to read

        break;
      }
    }
  }

  await waitFor(() => {
    expect(meta.controllers.size).toBe(0);
  });
  // wait for stopped
  await waitFor(() => {
    expect(stopped).toBe(true);
  });

  expect(yieldCalls.mock.calls.length).toBeGreaterThanOrEqual(3);
  expect(yieldCalls.mock.calls.length).toBeLessThan(10);

  const errors = onConsumerErrorSpy.mock.calls.map(
    (it) => (it[0].error as Error).message,
  );
  expect(errors).toMatchInlineSnapshot(`
    Array [
      "The operation was aborted.",
    ]
  `);
  expect(onConsumerErrorSpy).toHaveBeenCalledTimes(1);
  expect(onProducerErrorSpy).toHaveBeenCalledTimes(0);

  await server.close();
});

test('e2e, encode/decode - maxDepth', async () => {
  const onError = vi.fn<(...args: Parameters<ProducerOnError>) => void>();
  const data = {
    0: Promise.resolve({
      foo: 'bar',
      deferred: Promise.resolve(42),
    }),
  } as const;
  const stream = jsonlStreamProducer({
    data,
    serialize: SuperJSON.serialize,
    onError,
    maxDepth: 1,
  });

  const server = createServerForStream(stream, new AbortController());

  const ac = new AbortController();
  const res = await fetch(server.url, {
    signal: ac.signal,
  });
  const [head] = await jsonlStreamConsumer<typeof data>({
    from: res.body!,
    deserialize: SuperJSON.deserialize,
    abortController: ac,
  });

  {
    expect(head[0]).toBeInstanceOf(Promise);

    const value = await head[0];
    await expect(value.deferred).rejects.toMatchInlineSnapshot(
      `[Error: Received error from server]`,
    );
  }

  expect(onError).toHaveBeenCalledTimes(1);
  expect(onError.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "error": [Error: Max depth reached at path: 0.deferred],
          "path": Array [
            "0",
            "deferred",
          ],
        },
      ],
    ]
  `);

  await server.close();
});
