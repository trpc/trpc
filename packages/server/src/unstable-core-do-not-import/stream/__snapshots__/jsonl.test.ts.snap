// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encode/decode with superjson 1`] = `
Array [
  Object {
    "json": Object {
      "0": Array [
        Array [
          0,
        ],
        Array [
          null,
          0,
          0,
        ],
      ],
      "1": Array [
        Array [
          0,
        ],
        Array [
          null,
          0,
          1,
        ],
      ],
    },
  },
  Object {
    "json": Array [
      0,
      1,
      Array [
        Array [
          Object {
            "deferred": 0,
            "foo": Object {
              "bar": Object {
                "baz": "qux",
              },
            },
          },
        ],
        Array [
          "deferred",
          0,
          2,
        ],
      ],
    ],
  },
  Object {
    "json": Array [
      1,
      1,
      Array [
        Array [
          0,
        ],
        Array [
          null,
          1,
          3,
        ],
      ],
    ],
  },
  Object {
    "json": Array [
      2,
      1,
      Array [
        Array [
          42,
        ],
      ],
    ],
  },
  Object {
    "json": Array [
      3,
      1,
      Array [
        Array [
          1,
        ],
      ],
    ],
  },
  Object {
    "json": Array [
      3,
      1,
      Array [
        Array [
          2,
        ],
      ],
    ],
  },
  Object {
    "json": Array [
      3,
      1,
      Array [
        Array [
          3,
        ],
      ],
    ],
  },
  Object {
    "json": Array [
      3,
      2,
      Array [
        Array [],
      ],
    ],
  },
]
`;
